### PR TITLE
Use chroot for containerized mounts

### PR DESCRIPTION
**Pull request by jingxu97**
This PR is to modify the containerized mounter script to use chroot
instead of rkt fly. This will avoid the problem of possible large number
of mounts caused by rkt containers if they are not cleaned up.

### CODE REVIEW

**Comment by fejta** [comment]
Please assign to relevant people once tests pas

File: cluster/gce/gci/configure-helper.sh [hunk 1]
```diff
@@ -880,6 +880,18 @@ function compute-master-manifest-variables {
   fi
 }
 
+# A helper function that bind mounts kubelet dirs for running mount in a chroot
+function bind-mount-mounterRootfs {
```
Comment (vishh) [inline cluster/gce/gci/configure-helper.sh]: nit: I'd name it `prepare-mounter-rootfs` instead

File: cluster/gce/gci/configure-helper.sh [hunk 2]
```diff
@@ -880,6 +880,18 @@ function compute-master-manifest-variables {
...
+  echo "Prepare containerized mounter"
+  mount --bind "${CONTAINERIZED_MOUNTER_HOME}" "${CONTAINERIZED_MOUNTER_HOME}"
+  mount -o remount,exec "${CONTAINERIZED_MOUNTER_HOME}"
+  CONTAINERIZED_MOUNTER_ROOTFS="${CONTAINERIZED_MOUNTER_HOME}/rootfs"
+  mount --rbind /var/lib/kubelet/ "${CONTAINERIZED_MOUNTER_ROOTFS}/var/lib/kubelet"
+  mount --make-rshared "${CONTAINERIZED_MOUNTER_ROOTFS}/var/lib/kubelet"
+  mount --bind /proc "${CONTAINERIZED_MOUNTER_ROOTFS}/proc"
```
Comment (vishh) [inline cluster/gce/gci/configure-helper.sh]: recommend mounting a read only to be safe

File: cluster/gce/gci/configure.sh [hunk 3]
```diff
@@ -115,19 +115,18 @@ function split-commas {
...
+    chmod a+x "${CONTAINERIZED_MOUNTER_HOME}"
+    mkdir "${CONTAINERIZED_MOUNTER_HOME}/rootfs"
+    local -r mounter_tar_sha="8003b798cf33c7f91320cd6ee5cec4fa22244571"
+    download-or-bust "${mounter_tar_sha}" "https://storage.googleapis.com/kubernetes-release/gci-mounter/mounter.tar"
+    cp "${dst_dir}/kubernetes/gci-trusty/gci-mounter" "${CONTAINERIZED_MOUNTER_HOME}/mounter"
+    chmod a+x "${CONTAINERIZED_MOUNTER_HOME}/mounter"
+    mv "${KUBE_HOME}/mounter.tar" /tmp/mounter.tar
+    tar xvf /tmp/mounter.tar -C "${CONTAINERIZED_MOUNTER_HOME}/rootfs"
+    rm /tmp/mounter.tar
+    mkdir "${CONTAINERIZED_MOUNTER_HOME}/rootfs/var/lib/kubelet"
```
Comment (vishh) [inline cluster/gce/gci/configure.sh]: Maybe create `proc` and `dev` directories too?

File: cluster/gce/gci/mounter/mounter.go [hunk 4]
```diff
@@ -0,0 +1,93 @@
+/*
```
Comment (vishh) [inline cluster/gce/gci/mounter/mounter.go]: Ideally, we would maintain this mounter infrastructure outside of core k8s repo to avoid confusion. I'd recommend moving it out in a separate PR

File: cluster/gce/gci/mounter/mounter.go [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+Copyright 2016 The Kubernetes Authors.
```
Comment (vishh) [inline cluster/gce/gci/mounter/mounter.go]: nit: Change year

**Review by vishh — COMMENTED** [review]
Overall this PR lgtm. I did not do a thorough review though. May be someone else should do a more detailed review before merging it?

**Comment by vishh** [comment]
@msau42 interested in giving this PR a more detailed review?

**Comment by jsafrane** [comment]
lgtm for parts in `pkg/`. I did not check `cluster/gce`

File: cluster/gce/gci/configure.sh (hunk 3 above)
Comment (jingxu97) [inline cluster/gce/gci/configure.sh]: but there is already a /proc and /dev directory under /rootfs. Is this necessary?

File: cluster/gce/gci/mounter/mounter.go (hunk 4 above)
Comment (jingxu97) [inline cluster/gce/gci/mounter/mounter.go]: sure. I will do that.

**Comment by jingxu97** [comment]
@k8s-bot test this

File: cluster/gce/gci/configure-helper.sh [hunk 6]
```diff
@@ -880,6 +880,18 @@ function compute-master-manifest-variables {
   fi
 }
 
+# A helper function that bind mounts kubelet dirs for running mount in a chroot
+function prepare-mounter-rootfs {
+  echo "Prepare containerized mounter"
+  mount --bind "${CONTAINERIZED_MOUNTER_HOME}" "${CONTAINERIZED_MOUNTER_HOME}"
```
Comment (msau42) [inline cluster/gce/gci/configure-helper.sh]: What does it mean to bind mount the same directory?

File: cluster/gce/gci/configure.sh [hunk 7]
```diff
@@ -115,19 +115,18 @@ function split-commas {
...
-    download-or-bust "${rkt_stage1_fly_sha1}" "https://storage.googleapis.com/kubernetes-release/rkt/${rkt_version}/stage1-fly.aci"
-    download-or-bust "${gci_mounter_sha1}" "https://storage.googleapis.com/kubernetes-release/gci-mounter/gci-mounter-${gci_mounter_version}.aci"
-    local -r rkt_dst="${KUBE_HOME}/bin/"
-    mv "${KUBE_HOME}/rkt" "${rkt_dst}/rkt"
-    mv "${KUBE_HOME}/stage1-fly.aci" "${rkt_dst}/stage1-fly.aci"
-    mv "${KUBE_HOME}/gci-mounter-${gci_mounter_version}.aci" "${rkt_dst}/gci-mounter-${gci_mounter_version}.aci"
-    chmod a+x "${rkt_dst}/rkt"
+    CONTAINERIZED_MOUNTER_HOME="${KUBE_HOME}/containerized_mounter"
+    mkdir "${CONTAINERIZED_MOUNTER_HOME}"
+    chmod a+x "${CONTAINERIZED_MOUNTER_HOME}"
```
Comment (msau42) [inline cluster/gce/gci/configure.sh]: Is "all" permissions necessary?

File: test/e2e_node/jenkins/gci-init.yaml [hunk 8]
```diff
@@ -3,10 +3,15 @@
...
   - mount /tmp /tmp -o remount,exec,suid
   - usermod -a -G docker jenkins
-  - mkdir -p /home/kubernetes/bin/
-  - mount -B /home/kubernetes/bin /home/kubernetes/bin
-  - mount -B -o remount,exec /home/kubernetes/bin 
-  - wget https://storage.googleapis.com/kubernetes-release/rkt/v1.18.0/rkt -O /home/kubernetes/bin/rkt
-  - wget https://storage.googleapis.com/kubernetes-release/rkt/v1.18.0/stage1-fly.aci -O /home/kubernetes/bin/stage1-fly.aci
-  - wget https://storage.googleapis.com/kubernetes-release/gci-mounter/gci-mounter-v2.aci -O /home/kubernetes/bin/gci-mounter-v2.aci
-  - chmod a+x /home/kubernetes/bin/rkt
+  - mkdir -p /var/lib/kubelet
```
Comment (msau42) [inline test/e2e_node/jenkins/gci-init.yaml]: How come these commands need to be run here in addition to the configure scripts?

File: cluster/gce/gci/configure-helper.sh [hunk 9]
```diff
@@ -1514,4 +1521,5 @@ else
   fi
 fi
 reset-motd
+bind-mount-mounterRootfs
```
Comment (msau42) [inline cluster/gce/gci/configure-helper.sh]: Where is this function defined?

Comment (jingxu97) [inline cluster/gce/gci/configure-helper.sh]: oh, my mistake. I changed the function name and forgot here.

File: test/e2e_node/jenkins/gci-init.yaml (hunk 8 above)
Comment (jingxu97) [inline test/e2e_node/jenkins/gci-init.yaml]: there setup is only for node e2e tests, it only create one node which runs everything.

File: cluster/gce/gci/configure.sh (hunk 7 above)
Comment (jingxu97) [inline cluster/gce/gci/configure.sh]: I think u+x is fine. but put it a+x just in case.

File: cluster/gce/gci/configure-helper.sh (hunk 6 above)
Comment (jingxu97) [inline cluster/gce/gci/configure-helper.sh]: It is a copy-paste thing, I assumed it is for remount. But you are right, it is not necessary, I think

**Comment by jingxu97** [comment]
@k8s-bot gci gke e2e test this

**Comment by jingxu97** [comment]
@k8s-bot  test this

**Comment by jingxu97** [comment]
@msau42 @vishh  I addressed your comments, PTAL. Thanks!

**Comment by jingxu97** [comment]
This PR is ready to be reviewed. @msau42 @vishh @jsafrane  PTAL

File: cluster/gce/gci/configure-helper.sh (hunk 6 above)
Comment (vishh) [inline cluster/gce/gci/configure-helper.sh]: It is necessary for remount which can be performed only on existing mounts.

File: cluster/gce/gci/configure.sh (hunk 3 above)
Comment (vishh) [inline cluster/gce/gci/configure.sh]: That's an implicit assumption. What if it's not there. Maybe fix it in a separate PR?

**Review by vishh — APPROVED** [review]

**Comment by lucab** [comment]
> This PR is to modify the containerized mounter script to use chroot
instead of rkt fly. This will avoid the problem of possible large number
of mounts caused by rkt containers if they are not cleaned up.

rkt stage-fly is effectively a chroot, and the large number of mounts is caused by rshared propagation back to the host in case of spin-crashing kubelet (or any other process/container, see https://github.com/kubernetes/community/pull/193#issuecomment-273863723). This PR seems to be performing exactly the same steps, so I'm not completely sure how it avoids the problem.

**Comment by vishh** [comment]
@lucab Instead of creating multiple rkt fly pods, this PR is instead creating a single pod filesystem sandbox and re-using it multiple times. If rkt supported `rkt create` followed by multiple unrelated `rkt run`, then this PR could have used `rkt` instead.

File: cluster/gce/gci/configure.sh [hunk 10]
```diff
@@ -115,19 +115,18 @@ function split-commas {
...
+    mkdir "${CONTAINERIZED_MOUNTER_HOME}/rootfs"
+    local -r mounter_tar_sha="8003b798cf33c7f91320cd6ee5cec4fa22244571"
+    download-or-bust "${mounter_tar_sha}" "https://storage.googleapis.com/kubernetes-release/gci-mounter/mounter.tar"
+    cp "${dst_dir}/kubernetes/gci-trusty/gci-mounter" "${CONTAINERIZED_MOUNTER_HOME}/mounter"
+    chmod a+x "${CONTAINERIZED_MOUNTER_HOME}/mounter"
+    mv "${KUBE_HOME}/mounter.tar" /tmp/mounter.tar
+    tar xvf /tmp/mounter.tar -C "${CONTAINERIZED_MOUNTER_HOME}/rootfs"
```
Comment (mikedanese) [inline cluster/gce/gci/configure.sh]: This never succeeds. It also prints out >10,000 lines:

```
+ tar xf /tmp/mounter.tar -C /home/kubernetes/containerized_mounter/rootfs
tar: dev/agpgart: Cannot mknod: Read-only file system
tar: dev/audio: Cannot mknod: Read-only file system
tar: dev/audio1: Cannot mknod: Read-only file system
tar: dev/audio2: Cannot mknod: Read-only file system
tar: dev/audio3: Cannot mknod: Read-only file system
tar: dev/audioctl: Cannot mknod: Read-only file system
tar: dev/console: Cannot open: File exists
tar: dev/core: Cannot open: File exists
tar: dev/dsp: Cannot mknod: Read-only file system
tar: dev/dsp1: Cannot mknod: Read-only file system
tar: dev/dsp2: Cannot mknod: Read-only file system
tar: dev/dsp3: Cannot mknod: Read-only file system
tar: dev/fd: Cannot open: File exists
tar: dev/full: Cannot mknod: File exists
tar: dev/kmem: Cannot mknod: Read-only file system
tar: dev/loop0: Cannot mknod: File exists
tar: dev/loop1: Cannot mknod: File exists
tar: dev/loop2: Cannot mknod: File exists
tar: dev/loop3: Cannot mknod: File exists
tar: dev/loop4: Cannot mknod: File exists
tar: dev/loop5: Cannot mknod: File exists
...
```
